src/model_signing/manifest.py

```diff
@@ -245,16 +246,156 @@ def key(self) -> ManifestKey:
         return Shard(self._path, self._start, self._end)
 
 
+class SerializationType(metaclass=abc.ABCMeta):
+    """A description of the serialization process that generated the manifest.
+
+    These should record all the parameters needed to ensure a reproducible
+    serialization. These are used to build a manifest back from the signature in
+    a backwards compatible way. We use these to determine what serialization to
+    use when verifying a signature.
+    """
+
+    @property
+    @abc.abstractmethod
+    def serialization_dict(self) -> dict[str, Any]:
+        """The arguments of the serialization method."""
+
+    @classmethod
+    def from_dict(cls, serialization_dict: dict[str, Any]) -> Self:
+        """Builds an instance of this class based on the dict representation.
+
+        This is the reverse of `serialization_dict`.
+        """
+        serialization_type = serialization_dict["method"]
+        for subclass in [_FileSerialization, _ShardSerialization]:
+            if serialization_type == subclass.method:
+                return subclass._from_dict(serialization_dict)
+        raise ValueError(f"Unknown serialization type {serialization_type}")
+
+    @classmethod
+    @abc.abstractmethod
+    def _from_dict(cls, serialization_dict: dict[str, Any]) -> Self:
+        """Performs the actual build from `from_dict`."""
+
+    @abc.abstractmethod
+    def new_item(self, name: str, digest: hashing.Digest) -> ManifestItem:
+        """Returns a new manifest item with given name and digest."""
+
+
+class _FileSerialization(SerializationType):
+    method: Final[str] = "files"
+
+    def __init__(self, hash_type: str, allow_symlinks: bool = False):
+        """Records the manifest serialization type for serialization by files.
+
+        We only need to record the hashing engine used and whether symlinks are
+        hashed or ignored.
+
+        Args:
+            hash_type: A string representation of the hash algorithm.
+            allow_symlinks: Controls whether symbolic links are included.
+        """
+        self._hash_type = hash_type
+        self._allow_symlinks = allow_symlinks
+
+    @property
+    @override
+    def serialization_dict(self) -> dict[str, Any]:
+        return {
+            "method": self.method,
+            "hash_type": self._hash_type,
+            "allow_symlinks": self._allow_symlinks,
+        }
+
+    @classmethod
+    @override
+    def _from_dict(cls, serialization_dict: dict[str, Any]) -> Self:
+        return cls(
+            serialization_dict["hash_type"],
+            serialization_dict["allow_symlinks"],
+        )
+
+    @override
+    def new_item(self, name: str, digest: hashing.Digest) -> ManifestItem:
+        path = pathlib.PurePosixPath(name)
+        return FileManifestItem(path=path, digest=digest)
+
+
+class _ShardSerialization(SerializationType):
+    method: Final[str] = "shards"
```

Comment:
do we expect any other type of shards where we would want to distinguish this as a file shard?

Reply:
If we do, that would be a separate serializer, with separate manifest item type and separate serialization type. But right now we should only have these two types

Reply:
From a intoto standpoint I'm wondering about this part. Since the method here would need to be unique?
```
  "predicate": {
    "serialization": {
      "method": "shards",
```

Reply:
Yes, we will guarantee it to be unique